src/main/java/InfrastructureManager/Master.java

```diff
@@ -1,37 +1,138 @@
 package InfrastructureManager;
 
+import java.util.ArrayList;
+
+/**
+ * Master Class of the Infrastructure Manager, singleton class
+ */
 public class Master {
 
     private final CommandSet commandSet;
-    private final MasterInput input;
-    private final MasterOutput output;
+    private final ArrayList<Runner> runnerList;
+    private final ArrayList<Runner> running; //List to store currently running runners
+
+    private static Master instance = null;
 
-    public Master() {
+    /**
+     * Constructor of the class
+     * Gets the master configured according to the config file.
+     */
+    private Master() {
         MasterConfigurator configurator = new MasterConfigurator();
         commandSet = configurator.getCommands();
-        input =configurator.getInput();
-        output = configurator.getOutput();
-    }
-    public String fromInput() {
-        return this.input.read();
+        runnerList = configurator.getRunners();
+        running = new ArrayList<>();
     }
+
+    /**
+     * Based on the command set predefined, execute a command
+     * @param command Command to be executed coming from the input
+     * @return Response going to the output(s)
+     */
     public String execute(String command) {
         return this.commandSet.getResponse(command);
     }
-    public void toOutput(String response) {
-        this.output.out(response);
+
+    /**
+     * Method for exiting the program, by exiting each running runner
+     */
+    public void exitAll() {
+        for (Runner runner : running) {
+            runner.exit();
+        }
     }
 
-    public static void main(String[] args) {
-        Master master = new Master();
-        String in;
-        String mapping;
-        while (true) {
-            in = master.fromInput();
-            mapping = master.execute(in);
-            master.toOutput(mapping);
-            if (in.equals("exit")) {break;}
+    /**
+     * Method for starting the main runner thread, declared in the config file
+     */
+    //TODO:This could maybe go in configurator
+    public void startMainRunner() {
+        for (Runner runner : runnerList) {
+            if (runner.getName().equals("Main")) {
+                running.add(runner);
+                new Thread(runner,"MainRunner").start();
+                break;
+            }
+        }
+    }
+
+    /**
+     * Method that given a certain Scenario, runs the corresponding ScenarioRunner and adds
+     * it to the list of running Runners. If called on a running ScenarioRunner, it will restart
+     * it
+     * @param scenario Scenario to be run
+     */
+    public void runScenario(Scenario scenario) {
+        ScenarioRunner scenarioRunner = getRunner(scenario);
+        if (running.contains(scenarioRunner)) { //If running then stop it and re-run
```

Comment:
why? if running just leave it running.

Reply:
I was thinking about whats written in the issue #17 as a wanted feature for the ScenarioDispatcher:

> Run (or restart) a Scenario.

So, in this way, by calling run again on a scenario, it would restart.

Reply:
My concern is if we have two different inputs that try to start the same event, then the second one may unintentionally restart it, I think we need a stop function.
By the way, the design in the original issue is preliminary and may change on the way based on problems we face. The important part is that we don't "just fix the problem", instead we alter the design and make sure that the changes are reflected everywhere not just where the problem occurs.

Reply:
You are right, I havent considered that case. I will add the stop function, also in the dispatcher, and allow the running runners to just keep running. 👍